pkg/storegateway/series_chunks_test.go

```diff
@@ -515,7 +627,98 @@ func TestLoadingSeriesChunksSetIterator(t *testing.T) {
 			for _, s := range loadedSets {
 				s.release()
 			}
+
 			assert.Zero(t, int(bytesPool.balance.Load()))
+			assert.Zero(t, seriesEntrySlicePool.(*pool.TrackedPool).Balance.Load())
+			assert.Greater(t, seriesEntrySlicePool.(*pool.TrackedPool).Gets.Load(), int64(0))
+			assert.Zero(t, seriesChunksSlicePool.(*pool.TrackedPool).Balance.Load())
+			assert.Greater(t, seriesChunksSlicePool.(*pool.TrackedPool).Gets.Load(), int64(0))
+		})
+	}
+}
+
+func BenchmarkLoadingSeriesChunksSetIterator(b *testing.B) {
+	for batchSize := 10; batchSize <= 10000; batchSize *= 10 {
+		b.Run(fmt.Sprintf("batch size: %d", batchSize), func(b *testing.B) {
+			var (
+				numSets            = 100
+				numSeriesPerSet    = batchSize
+				numChunksPerSeries = 10
+			)
+
+			// For simplicity all series have the same chunks. This doesn't affect the benchmark anyway.
+			blockID := ulid.MustNew(1, nil)
+			seriesEntries := make([]seriesEntry, 0, numChunksPerSeries)
+			for chunkIdx := 0; chunkIdx < numChunksPerSeries; chunkIdx++ {
+				seriesEntries = append(seriesEntries, seriesEntry{
+					refs: []chunks.ChunkRef{chunks.ChunkRef(chunkIdx)},
+					chks: []storepb.AggrChunk{{MinTime: int64(chunkIdx), MaxTime: int64(chunkIdx), Raw: &storepb.Chunk{}}},
+				})
+			}
+
+			// Creates the series sets, guaranteeing series sorted by labels.
+			sets := make([]seriesChunkRefsSet, 0, numSets)
+			for setIdx := 0; setIdx < numSets; setIdx++ {
+				minSeriesID := (numSets * numSeriesPerSet) + (setIdx * numSeriesPerSet)
+				maxSeriesID := minSeriesID + numSeriesPerSet - 1
+
+				// This set cannot be released because reused between multiple benchmark runs.
+				set := createSeriesChunkRefsSet(minSeriesID, maxSeriesID, false)
+
+				for seriesIdx := 0; seriesIdx < set.len(); seriesIdx++ {
+					set.series[seriesIdx].chunks = make([]seriesChunkRef, numChunksPerSeries)
+					for chunkIdx := 0; chunkIdx < numChunksPerSeries; chunkIdx++ {
+						set.series[seriesIdx].chunks[chunkIdx] = seriesChunkRef{
+							blockID: blockID,
+							ref:     chunks.ChunkRef(chunkIdx),
+						}
+					}
```

Comment:
nit: there is `generateSeriesChunkRef`, only difference i see is that it has a different block ULID for each chunk

Reply:
Good suggestion. Here I need the control the blockID, but other tests don't need a different blockID for each chunk, so I've changed the function and reused it here: https://github.com/grafana/mimir/pull/3739/commits/9fa044cda803d7db6e1042687a1b9c1708aa3b97

Reply:
oh i hadn't seen the chunkReaders. Thanks for addressing this